[Merged by Bors] - feat(RingTheory): define R-linear graded algebra homomorphism

kckennylau opened the pull request:
This PR defines R-linear graded algebra homomorphisms, which are R-algebra homomorphisms that preserve the grading. We also provide the basic properties (aka the "API").

Zulip discussion: [#mathlib4 > redefine Graded Algebra](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/redefine.20Graded.20Algebra/near/543847326)

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

Split from #26061.

- [x] depends on: #30355 
- [x] depends on: #30312 

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

kckennylau committed:
feat(RingTheory): define graded ring homomorphisms

kckennylau committed:
mk_all

kckennylau committed:
remove invisible character

kckennylau committed:
lint

kckennylau committed:
Synchronise with #30334

kckennylau committed:
make GradedRingHom extend FunLike (sync #30334)

kckennylau committed:
graded alg hom draft

kckennylau committed:
feat(Logic): graded functions

kckennylau committed:
mk_all

kckennylau committed:
add module docstring

kckennylau committed:
Merge remote-tracking branch 'fork/graded-funlike' into graded-ring-hom

kckennylau committed:
refactor GradedRingHomClass into GradedFunLike and RingHomClass

kckennylau committed:
Merge remote-tracking branch 'fork/graded-ring-hom' into graded-alg-hom

kckennylau committed:
feat(RingTheory): define R-linear graded algebra homomorphism

kckennylau committed:
remove unused arguments

kckennylau committed:
update

kckennylau committed:
update

kckennylau committed:
remove dead comment

erdOne commented on Mathlib/RingTheory/GradedAlgebra/AlgHom.lean:
```diff
@@ -0,0 +1,268 @@
+`[FunLike F A B] [GradedFunLike F 𝒜 ℬ] [AlgHomClass F R A B]` into an actual `GradedAlgHom`.
+
+In future mathlib this will be deprioritised in favour of using structural projections. -/
+@[coe]
```
This shouldn't be coe/simpNF. See the recent discussion in Zulip.

erdOne reviewed (commented):
Otherwise LGTM

Thanks!
maintainer delegate

kckennylau committed:
address comments

kckennylau committed:
fix error

kckennylau committed:
address comments

kckennylau committed:
Update AlgHom.lean

kckennylau committed:
Rename toAlgHom to toEnd and enrich as MonoidHom

Enrich the `toAlgHom` function with `MonoidHom` structure, renaming it to `toEnd`.
